haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -229,45 +311,387 @@ def run(  # noqa: PLR0915, PLR0912
+        self.debug_path.mkdir(exist_ok=True)
+
+        dt = datetime.now()
+        file_name = Path(f"breakpoint_at_{component_name}_{dt.strftime('%Y_%m_%d_%H_%M_%S')}.json")
+        state = {
+            "input_data": self._serialize_component_input(self.original_input_data),  # original input data
+            "timestamp": dt.isoformat(),
+            "breakpoint": {"component": component_name, "visits": component_visits[component_name]},
+            "pipeline_state": {
+                "inputs": self._serialize_component_input(inputs),  # current pipeline state inputs
+                "component_visits": component_visits,
+                "ordered_component_names": self.ordered_component_names,
+            },
+        }
+
+        try:
+            with open(self.debug_path / file_name, "w") as f_out:
+                json.dump(state, f_out, indent=2)
+            logger.info(f"Pipeline state saved at: {file_name}")
+
+            # pass the state to some user-defined callback function
+            if callback_fun is not None:
+                callback_fun(state)
+
+            return state
+
+        except Exception as e:
+            logger.error(f"Failed to save pipeline state: {str(e)}")
+            raise
+
+    @staticmethod
+    def load_state(file_path: Union[str, Path]) -> Dict[str, Any]:
+        """
+        Load a saved pipeline state.
+
+        :param file_path: Path to the state file
+        :returns:
+            Dict containing the loaded state
+        """
+        import json
+
+        file_path = Path(file_path)
+
+        try:
+            with open(file_path, "r", encoding="utf-8") as f:
+                state = json.load(f)
+            Pipeline._validate_resume_state(state=state)
+            logger.info(f"Successfully loaded pipeline state from: {file_path}")
+            return state
+        except FileNotFoundError:
+            raise FileNotFoundError(f"File not found: {file_path}")
+        except json.JSONDecodeError as e:
+            raise json.JSONDecodeError(f"Invalid JSON file {file_path}: {str(e)}", e.doc, e.pos)
+        except IOError as e:
+            raise IOError(f"Error reading {file_path}: {str(e)}")
+
+    def _validate_pipeline_state(self, resume_state: Dict[str, Any]) -> None:
+        """
+        Validates that the resume_state contains valid configuration for the current pipeline.
+
+        Raises a PipelineRuntimeError if any component is missing or if the state structure is invalid.
+
+        :param resume_state: The saved state to validate.
+        """
+
+        pipeline_state = resume_state["pipeline_state"]
+        valid_components = set(self.graph.nodes.keys())
+
+        # Check if the ordered_component_names are valid components in the pipeline
+        missing_ordered = set(pipeline_state["ordered_component_names"]) - valid_components
+        if missing_ordered:
+            raise PipelineRuntimeError(
```

Comment:
As a heads up `PipelineRuntimeError` recently got refactored in this [PR](https://github.com/deepset-ai/haystack/pull/9182) which could change how you call it.